### PR TITLE
Check that the string isn't empty

### LINKED ISSUE #1819: assertion failure due to empty string in XmpTextValue::read()
The fix in #1769 was incomplete. The problem is this code in [value.cpp](https://github.com/Exiv2/exiv2/blob/c65941c4cd0f1e3382367105eaf30097188d3f40/src/value.cpp#L627-L632):

```c
if (type.empty()) {
    throw Error(kerInvalidXmpText, type);
}
// Strip quotes (so you can also specify the type without quotes)
if (type[0] == '"') type = type.substr(1);
if (type[type.length()-1] == '"') type = type.substr(0, type.length()-1);
```

The `type.empty()` check can be invalidated if the next character is a `"`.

This bug is reproducible on `main` and `0.27-maintenance`:

```bash
$ exiv2 poc
/usr/include/c++/10/bits/basic_string.h:1038: std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::reference std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::operator[](std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::size_type) [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>; std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::reference = char&; std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::size_type = long unsigned int]: Assertion '__pos <= size()' failed.
Aborted (core dumped)
```

poc: [poc.tar.gz](https://github.com/Exiv2/exiv2/files/6911307/poc.tar.gz)

### PR DESCRIPTION
**Pull request by kevinbackhouse**
Fixes: #1819

There were a couple of places where we checked that string wasn't empty, but then removed another character, thereby sometimes making the string empty again.

The output of `test_issue_ghsa_v5g7_46xf_h728.py` changed because I removed [this check](https://github.com/Exiv2/exiv2/blob/c65941c4cd0f1e3382367105eaf30097188d3f40/src/value.cpp#L628) and made it safe in a different way.

### CODE REVIEW

File: src/value.cpp [hunk 1]
```diff
@@ -697,12 +695,9 @@ namespace Exiv2 {
...
             std::string::size_type pos = buf.find_first_of(' ');
             type = buf.substr(5, pos-5);
-            if (type.empty()) {
-                throw Error(kerInvalidXmpText, type);
-            }
             // Strip quotes (so you can also specify the type without quotes)
-            if (type[0] == '"') type = type.substr(1);
-            if (type[type.length()-1] == '"') type = type.substr(0, type.length()-1);
+            if (!type.empty() && type[0] == '"') type = type.substr(1);
+            if (!type.empty() && type[type.length()-1] == '"') type = type.substr(0, type.length()-1);
```
Comment (hassec) [inline src/value.cpp:700]:
```suggestion
            if (!type.empty() && type.back() == '"') type.pop_back();
```

File: src/value.cpp [hunk 2]
```diff
@@ -488,13 +488,11 @@ namespace Exiv2 {
...
-            std::string::size_type pos = comment.find_first_of(' ');
+            const std::string::size_type pos = comment.find_first_of(' ');
             std::string name = comment.substr(8, pos-8);
             // Strip quotes (so you can also specify the charset without quotes)
-            if (!name.empty()) {
-                if (name[0] == '"') name = name.substr(1);
-                if (name[name.length()-1] == '"') name = name.substr(0, name.length()-1);
-            }
+            if (!name.empty() && name[0] == '"') name = name.substr(1);
+            if (!name.empty() && name[name.length()-1] == '"') name = name.substr(0, name.length()-1);
```
Comment (hassec) [inline src/value.cpp:495]:
```suggestion
            if (!name.empty() && name.back() == '"') name.pop_back();
```

File: src/value.cpp [hunk 3]
```diff
@@ -488,13 +488,11 @@ namespace Exiv2 {
         std::string c = comment;
         CharsetId charsetId = undefined;
         if (comment.length() > 8 && comment.substr(0, 8) == "charset=") {
-            std::string::size_type pos = comment.find_first_of(' ');
+            const std::string::size_type pos = comment.find_first_of(' ');
```
Comment (hassec) [inline src/value.cpp:491]: if we are touching this line anyway, WDYT about:
```suggestion
            const auto pos = comment.find_first_of(' ');
```

File: src/value.cpp [hunk 4]
```diff
@@ -862,14 +857,18 @@ namespace Exiv2 {
...
+            } else {
+                lang = buf.substr(5, pos-5);
+            }
             if (lang.empty()) throw Error(kerInvalidLangAltValue, buf);
             // Strip quotes (so you can also specify the language without quotes)
             if (lang[0] == '"') {
                 lang = lang.substr(1);
 
-                if (lang == "" || lang.find('"') != lang.length()-1)
+                if (lang.empty() || lang.find('"') != lang.length()-1)
```
Comment (hassec) [inline src/value.cpp:871]:
```suggestion
                if (lang.empty() || '"' != lang.back())
```

**Review by hassec — CHANGES_REQUESTED** [review]

File: src/value.cpp (hunk 3 above)
Comment (kevinbackhouse) [reply src/value.cpp:491]: I think that might be a C++11 feature that we can't use on 0.27-maintenance.

File: src/value.cpp (hunk 4 above)
Comment (kevinbackhouse) [reply src/value.cpp:871]: That would change the behavior slightly, although your version might be what was actually intended. The old code also throws an exception if there's a `"` character in the middle of the string. Which behavior do you think we want?

File: src/value.cpp (hunk 3 above)
Comment (hassec) [reply src/value.cpp:491]: Sorry @kevinbackhouse :see_no_evil:  I forgot to take into account that this is of course only C++98.  
But that's also bad news for my other comments `back()` and `pop_back` are both C++11 features....

Sorry again!

File: src/value.cpp (hunk 4 above)
Comment (hassec) [reply src/value.cpp:871]: see comment above, we can't use `back()` anyway, so let's leave this as is.

File: src/value.cpp (hunk 3 above)
Comment (kevinbackhouse) [reply src/value.cpp:491]: Good catch! I didn't realize that `back` and `pop_back` are C++11 too. I wonder why the build didn't break? I'll do a force push to revert these changes.

**Review by hassec — APPROVED** [review]

**Pull request merged**